### PR TITLE
Marshaling helpers for MarshalAs LPUTF8Str.

### PR DESCRIPTION
**Pull request by tijoytom**
Use the same Encoding.UTF8 APIs that we use
to marshal.These new APIs will go to S.R.I.PInvoke
contracts under PInvokeMarshal class.

### CODE REVIEW

**Comment by tijoytom** [comment]
@yizhang82 @RyanByi @weshaggard 
Marshaling helpers for UTF8

File: src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs [hunk 1]
```diff
@@ -1872,6 +1900,38 @@ unsafe public static IntPtr StringToCoTaskMemUni(String s)
...
+        unsafe public static IntPtr StringToAllocatedMemoryUTF8(String s)
+        {
+            const int MAX_UTF8_CHAR_SIZE = 3;
+            if (s == null)
+            {
+                return IntPtr.Zero;
+            }
+            else
+            {
+                int nb = (s.Length + 1) * MAX_UTF8_CHAR_SIZE;
```
Comment (tijoytom) [inline src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs]: We are probably allocating more than required here , we can allocate the correct size using GetByteCount

File: src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs [hunk 2]
```diff
@@ -185,7 +185,40 @@ public static String PtrToStringAuto(IntPtr ptr)
...
+            int nbBytes = System.StubHelpers.StubHelpers.strlen((sbyte*)ptr.ToPointer());
+            return PtrToStringUTF8(ptr, nbBytes);
+        }
+
+        [System.Security.SecurityCritical]  // auto-generated_required
+        unsafe public static String PtrToStringUTF8(IntPtr ptr,int byteLen)
+        {
+            if (byteLen < 0)
+            {
+                throw new ArgumentException("len");
```
Comment (yizhang82) [inline src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs]: byteLen

File: src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs [hunk 3]
```diff
@@ -1060,7 +1093,7 @@ public static IntPtr GetHINSTANCE(Module m)
             }
 
             if (rtModule == null)
-                throw new ArgumentNullException("m", Environment.GetResourceString("Argument_MustBeRuntimeModule"));
+                throw new ArgumentNullException(Environment.GetResourceString("Argument_MustBeRuntimeModule"));
```
Comment (yizhang82) [inline src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs]: Why are you removing "m"?

File: src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs [hunk 4]
```diff
@@ -1872,6 +1905,38 @@ unsafe public static IntPtr StringToCoTaskMemUni(String s)
         }
 
         [System.Security.SecurityCritical]  // auto-generated_required
+        unsafe public static IntPtr StringToAllocatedMemoryUTF8(String s)
```
Comment (yizhang82) [inline src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs]: Unfortunately, given that we no longer have the Pinvoke contract, we need to name this StringToCoTaskMemUTF8.

**Comment by yizhang82** [comment]
@tijoytom LGTM, but please do address the feedback in a future change before RTM.

File: src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs (hunk 3 above)
Comment (tijoytom) [inline src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs]: @yizhang82 that's weird , it's merge conflict and git just over-wrote it and did not see it as a merge conflict. This change to use "m" was from the previous PR ( Hugh Bellamy)

File: src/mscorlib/model.xml [hunk 5]
```diff
@@ -5915,6 +5915,8 @@
       <Member Name="PrelinkAll(System.Type)" />
       <Member Name="PtrToStringAnsi(System.IntPtr)" />
       <Member Name="PtrToStringAnsi(System.IntPtr,System.Int32)" />
+      <Member Name="PtrToStringUTF8(System.IntPtr)" />
```
Comment (weshaggard) [inline src/mscorlib/model.xml]: Please add these to System.Private.CoreLib model.xml as well.
cc @gkhanna79 @AlexGhiondea

Comment (gkhanna79) [inline src/mscorlib/model.xml]: @tijoytom Can you please look into fixing this ASAP as it will block RTM.

**Pull request merged**